src/app/clusters/commodity-tariff-server/CommodityTariffAttrsDataMgmt.cpp

```diff
@@ -0,0 +1,817 @@
+                       const CommodityTariff::Structs::PeakPeriodStruct::Type & rhs)
+{
+    return (lhs.severity == rhs.severity) && (lhs.peakPeriod == rhs.peakPeriod);
+}
+
+inline bool operator!=(const CommodityTariff::Structs::PeakPeriodStruct::Type & lhs,
+                       const CommodityTariff::Structs::PeakPeriodStruct::Type & rhs)
+{
+    return !(lhs == rhs);
+}
+
+} // namespace Clusters
+
+} // namespace app
+} // namespace chip
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::Platform;
+using namespace chip::app::DataModel;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::Globals;
+using namespace chip::app::Clusters::Globals::Structs;
+using namespace chip::app::Clusters::CommodityTariff;
+using namespace chip::app::Clusters::CommodityTariff::Structs;
+
+// static constexpr size_t kAuxSwitchesSettingsMax = 8;
+
+#define VerifyOrReturnError_LogSend(expr, code, ...)                                                                               \
+    do                                                                                                                             \
+    {                                                                                                                              \
+        ChipLogError(NotSpecified, __VA_ARGS__);                                                                                   \
+        VerifyOrReturnError(expr, code);                                                                                           \
+    } while (false)
+
+namespace CommonUtilities {
+static bool HasFeatureInCtx(TariffUpdateCtx * aCtx, Feature aFeature)
+{
+    return aCtx->mFeature.Has(aFeature);
+}
+
+static void CleanUpIDs(DataModel::List<const uint32_t> & IDs)
+{
+    if (!IDs.empty() && IDs.data())
+    {
+        MemoryFree(const_cast<uint32_t *>(IDs.data()));
+        IDs = DataModel::List<const uint32_t>();
+    }
+}
+
+static bool HasDuplicateIDs(const DataModel::List<const uint32_t> & IDs, std::unordered_set<uint32_t> & seen)
+{
+    for (auto id : IDs)
+    {
+        if (!seen.insert(id).second)
+        {
+            return true; // Duplicate found
+        }
+    }
+    return false;
+}
+
+template <typename T>
+static bool AreOptionalNullableEqual(const Optional<DataModel::Nullable<T>> & lhs, const Optional<DataModel::Nullable<T>> & rhs)
+{
+    if (lhs.HasValue() != rhs.HasValue())
+        return false;
+    if (!lhs.HasValue())
+        return true;
+    if (lhs.Value().IsNull() != rhs.Value().IsNull())
+        return false;
+    if (lhs.Value().IsNull())
+        return true;
+    return (lhs.Value().Value() != rhs.Value().Value());
+}
+
+template <typename T>
+static bool AreOptionalEqual(const Optional<T> & lhs, const Optional<T> & rhs)
+{
+    if (lhs.HasValue() != rhs.HasValue())
+        return false;
+    if (!lhs.HasValue())
+        return true;
+    return (lhs.Value() == rhs.Value());
```

Comment:
This seems like a really complicated way to write:
```suggestion
    return lhs == rhs;
```

But note that this will do the wrong thing for optional strings or optional lists...